README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

[![R-CMD-check](https://github.com/publichealthengland/fingertipscharts/workflows/R-CMD-check/badge.svg)](https://github.com/publichealthengland/fingertipscharts/actions)
[![codecov](https://codecov.io/gh/publichealthengland/fingertipscharts/branch/master/graph/badge.svg?token=D2X7WL46CC)](https://codecov.io/gh/publichealthengland/fingertipscharts)

# fingertipscharts

This is an R package to help users to easily reproduce charts that are
displayed on Office for Health Improvement and Disparities’
[Fingertips](http://fingertips.phe.org.uk/) data tool. Along with the
`fingertipsR` package, this package can be used to help users bring the
data on the website into their own outputs.

## Installation

### With remotes

You can install the latest development version from github using
[remotes](https://github.com/r-lib/remotes):

``` r
# install.packages("remotes")
remotes::install_github("PublicHealthEngland/fingertipscharts",
                        build_vignettes = TRUE)
```

## Example of some visualisations

Here are some examples of visualisations the package provides. These are
expanded in the packages vignette.

### Trends

``` r
library(fingertipscharts)
library(dplyr)

# Create the data to plot
df <- create_test_data() %>%
        arrange(IndicatorName) %>%
        mutate(Timeperiod = rep(c("2011", "2012", "2013", "2014", "2015", "2016"),
                                each = 111))
country_val <- df %>%
  filter(AreaCode == "C001") %>%
  select(Timeperiod, Country_val = Value)

# add the signifance for the local area points compared to the comparator
df <- df %>%
  left_join(country_val, by = "Timeperiod") %>%
  mutate(Significance = case_when(
    LCI > Country_val ~ "Higher",
    UCI < Country_val ~ "Lower",
    TRUE ~ "Similar"
  ))

# plot the trends

p <- trends(df,
            timeperiod = Timeperiod,
            value = Value,
            area = AreaCode,
            comparator = "C001",
            area_name = "AC103",
            fill = Significance,
            lowerci = LCI,
            upperci = UCI,
            title = "Title of graph",
            subtitle = "AC103 compared to C001",
            xlab = "Year",
            ylab = "Unit of measurement")
p
```

![](man/figures/README-trends-1.png)<!-- -->

### Compare indicators

``` r
# Create the data to plot, filtering for areas within a single parent, and also the parent and the national area
region <- "PAC10"
top_names <- c("C001", region)
df <- create_test_data() %>%
  filter(IndicatorName == "Indicator 3",
         (ParentAreaCode == region |
            AreaCode %in% top_names))

# order the factor for the significance field so they appear in the legend in the desired order
ordered_levels <- c("Better",
                    "Similar", 
                    "Worse",
                    "Not compared")
df <- df %>%
        mutate(Significance = 
                       factor(Significance,
                              levels = ordered_levels))

# plot compare areas chart
p <- compare_areas(df,
                   AreaCode, 
                   Value,
                   fill = Significance,
                   lowerci = LCI,
                   upperci = UCI,
                   order = "desc",
                   top_areas = top_names,
                   title = unique(df$IndicatorName))
p
```

![](man/figures/README-compare-indicators-1.png)<!-- -->

# Area profiles

``` r
dfspine <- create_test_data()
p <- area_profiles(dfspine,
                   value = Value,
                   count = Count,
                   area_code = AreaCode,
                   local_area_code = "AC122",
                   indicator = IndicatorName,
                   timeperiod = Timeperiod,
                   trend = Trend,
                   polarity = Polarity,
                   significance = Significance,
                   area_type = AreaType,
                   median_line_area_code = "C001",
                   comparator_area_code = "PAC12",
                   datatable = TRUE,
                   relative_domain_text_size = 0.75,
                   relative_text_size = 1.2,
                   bar_width = 0.68,
                   indicator_label_nudgex = -0.1,
                   show_dividers = "outer",
                   header_positions = c(-1, -0.7, -0.44, -0.35, -0.25,
                       -0.15, -0.05, 1.08))
p
```

<img src="man/figures/README-area_profiles-1.png" width="100%" />


NEWS.md
# fingertipscharts 0.0.14

* fixed bug with map function caused by change in name of variables long and lat from ons api


# fingertipscharts 0.0.13

* fixed bug with map function
* removed dependency on `geojsonio`

# fingertipscharts 0.0.12

* Improved documentation for all charting functions
* Fixed interactive map where fill is not a factor
* Improved messaging for area_profiles() data issues
* Updated tests to testthat edition 3
* Adapted non-standard evaluation of code to use dplyr 1.0.0 sytax
* Graceful failure for API resource when unavailable so as not to impact CRAN tests

# fingertipscharts 0.0.11 (2020-06-05)

* trend argument added to `area_profiles()`
* `trends()` function now displays the area and comparator in the key for the chart
* slight amendment to `compare_areas()` and `map()` functions giving users flexibility to provide their own legend titles using `+ labs(fill = "Blah")` after function (see [issue 16](https://github.com/PublicHealthEngland/fingertipscharts/issues/16))
* `overview()` function now has an argument for `legend_position` whose ordering is controlled by the levels of the field assigned to `fill`
* fixed bug to allow users to modify the significance colours on spine chart

# fingertipscharts 0.0.10 (2019-10-07)

* Unnecessary warning messages for area_profiles() removed
* Fixed bug around ordering of factors filling the interactive maps

# fingertipscharts 0.0.9 (2019-09-09)

* Added example to `area_profiles()` to show how to specify the order of indicators and how to show domains
* Graceful fail if ons_api provided to map() function isn't responding


# fingertipscharts 0.0.8 (2019-07-18)

* No updates from a user perspective

# fingertipscharts 0.0.7 

* No updates from a user perspective

# fingertipscharts 0.0.6 (2019-05-13)

* `display.values` argument added to `compare_areas()` function - allowing users to display the values alongside the bars in the chart

# fingertipscharts 0.0.5 (2019-04-08)

* `area_profiles()` - users have control on number of decimal places to display non-count values at (including different decimal places for different indicators)

# fingertipscharts 0.0.4 (2019-02-07)

* added copyright_year argument to `map` function
* improved code coverage
* `area_profiles()` now should display the "Best/Highest" label better

# fingertipscharts 0.0.3 (2018-12-12)

* Improved functionality for `area_profiles()` to accept field names other than `IndicatorName` and `Polarity` for those two arguments

# fingertipscharts 0.0.2 (2018-09-14)

* Added detail around the `area_profiles()` function
* Added `create_test_data()`


# fingertipscharts 0.0.1

Due to popular demand, `fingertipscharts` has been created to help users create the charts they see on Public Health England's [Fingertips](https://fingertips.phe.org.uk/) website. There is an accompanying vignette that shows users the potential of the functions `browseVignettes("fingertipscharts")`


tests/a-l.R
library(testthat)
library(fingertipscharts)
library(dplyr)

test_check("fingertipscharts", filter = "^[a-l]")


tests/m-z.R
library(testthat)
library(fingertipscharts)
library(dplyr)

test_check("fingertipscharts", filter = "^[m-z]")


tests/testthat.R
library(testthat)
library(fingertipscharts)
test_check("fingertipscharts")


R/dummy_data.R
#' Create test data
#'
#' Create dataset to be used in tests (useful for demo purposes too)
#' @return a dummy data.frame of data
#' @import dplyr
#' @importFrom tidyselect where
#' @examples
#' df <- create_test_data()
#' @export
create_test_data <- function() {

        if (R.Version()$major >= 3 & R.Version()$minor >= 6.0) {
                suppressWarnings(set.seed(42, sample.kind = "Rounding"))
        } else {
                set.seed(42)
        }

        df_local <- data.frame(
            AreaCode = rep(paste0("AC", 100:199), 6),
            ParentAreaCode = rep(
                rep(
                    paste0("PAC", 10:19), 6),
                each = 10),
            IndicatorName = rep(
                paste("Indicator", 1:6),
                each = 100),
            Timeperiod = rep(c("2016", "2015/16", "2014 - 2016",
                               "Aug 2014", "2017 Q1", "2011 - 2016"),
                             each = 100),
            Polarity = rep(c("Not applicable      ", "RAG - Low is good   ",
                             "RAG - Low is good   ", "RAG - High is good  ",
                             "BOB - Blue orange blue", "RAG - High is good  "),
                           each = 100),
            AreaType = "Local",
            Count = sample(20:500, 600,
                           replace = T),
            Denominator = sample(500:1000, 600,
                                 replace = T)
        ) %>%
                mutate(
                    Value = 100 * .data$Count / .data$Denominator,
                    LCI = .data$Value * 0.95,
                    UCI = .data$Value * 1.05
                ) %>%
                mutate(across(tidyselect::where(is.factor),
                              as.character))

        df_parent <- df_local %>%
                group_by(.data$IndicatorName,
                         .data$ParentAreaCode,
                         .data$Timeperiod,
                         .data$Polarity) %>%
                summarise(Count = sum(.data$Count),
                          Denominator = sum(.data$Denominator),
                          .groups = "drop") %>%
                mutate(
                    Value = 100 * .data$Count / .data$Denominator,
                    LCI = .data$Value * 0.95,
                    UCI = .data$Value * 1.05,
                    AreaType = "Parent",
                    AreaCode = .data$ParentAreaCode,
                    ParentAreaCode = "C001"
                )

        df_country <- df_parent %>%
                group_by(.data$IndicatorName,
                         .data$ParentAreaCode,
                         .data$Timeperiod,
                         .data$Polarity) %>%
                summarise(Count = sum(.data$Count),
                          Denominator = sum(.data$Denominator),
                          .groups = "drop") %>%
                mutate(
                    Value = 100 * .data$Count / .data$Denominator,
                    LCI = .data$Value * 0.95,
                    UCI = .data$Value * 1.05,
                    AreaType = "Country",
                    AreaCode = .data$ParentAreaCode,
                    ParentAreaCode = NA,
                    Significance = "Not compared"
                )

        country_values <- df_country %>%
                select("IndicatorName",
                       EngVal = "Value")

        df_local <- df_local %>%
                left_join(country_values, by = "IndicatorName") %>%
                mutate(Significance = case_when(
                    grepl("^Not", .data$Polarity) ~ "None",
                    grepl("^RAG", .data$Polarity) ~ case_when(
                        (grepl("Low", .data$Polarity) &
                            .data$UCI < .data$EngVal) ~ "Better",
                        (grepl("Low", .data$Polarity) &
                             .data$LCI > .data$EngVal) ~ "Worse",
                        (grepl("High", .data$Polarity) &
                             .data$UCI < .data$EngVal) ~ "Worse",
                        (grepl("High", .data$Polarity) &
                             .data$LCI > .data$EngVal) ~ "Better",
                        TRUE ~ "Similar"),
                    .data$UCI < .data$EngVal ~ "Lower",
                    .data$LCI > .data$EngVal ~ "Higher",
                    TRUE ~ "Similar")) %>%
                select(!c("EngVal"))
        df_parent <- df_parent %>%
                left_join(country_values, by = "IndicatorName") %>%
                mutate(Significance = ifelse(grepl("^Not", .data$Polarity), "None",
                                             ifelse(grepl("^RAG", .data$Polarity),
                                                    ifelse(grepl("Low", .data$Polarity),
                                                           ifelse(.data$UCI < .data$EngVal, "Better",
                                                                  ifelse(.data$LCI > .data$EngVal, "Worse",
                                                                         "Similar")),
                                                           ifelse(.data$UCI < .data$EngVal, "Worse",
                                                                  ifelse(.data$LCI > .data$EngVal, "Better",
                                                                         "Similar"))),
                                                    ifelse(.data$UCI < .data$EngVal, "Lower",
                                                           ifelse(.data$LCI > .data$EngVal, "Higher",
                                                                  "Similar"))))) %>%
                select(!c("EngVal"))

        trend_categories <- c("Increasing and getting better",
                              "Increasing and getting worse",
                              "Decreasing and getting better",
                              "Decreasing and getting worse",
                              "No significant change",
                              "Could not be calculated",
                              "Increasing",
                              "Decreasing")
        df <- bind_rows(df_local,
                        df_parent,
                        df_country) %>%
                mutate(Domain = case_when(
                        grepl("1$", .data$IndicatorName) ~ "Dom 1",
                        grepl("2$", .data$IndicatorName) ~ "Dom 2",
                        TRUE ~ "Dom 3"),
                       Trend = sample(trend_categories, n(), replace = TRUE))
        return(df)

}


R/spine_chart_utils.R
#' Preprocess  data for spine chart
#'
#' Returns a data frame with the latest time period of data for each indicator
#' name.
#' @details This processing only takes place on the indicator field and the time
#'   period field provided. If the data contains multiple sexes or age groups
#'   for an indicator, make sure the indicator field reflects this.
#' @inheritParams area_profiles
#' @param indicator unquoted field name for indicators. This should be what is
#'   presented as the label for the final spine chart, hence should be unique
#'   for each vertabra. Be careful the indicator doesn't have sub-categories
#'   based on other fields, such as sex (male, female, persons) or age group
#' @param timeperiod_sortable unquoted field name containing the time period
#'   that is numeric and sortable, such that higher values are a later time
#'   period
#' @import dplyr
#' @return A processed data frame for latest time periods of given indicators
#' @export
spine_preprocess <- function(data, indicator, timeperiod_sortable) {
        # timeperiod_sortable <- enquo(timeperiod_sortable)
        # indicator <- enquo(indicator)
        data <- data %>%
                group_by({{ indicator }}) %>%
                filter({{ timeperiod_sortable }} ==
                               max({{ timeperiod_sortable }})) %>%
                ungroup()
}

#' Check function for multiple values for an area in an indicator for spine chart
#' @inheritParams area_profiles
spine_data_check <- function(data, indicator, area_code) {
        data <- data %>%
                count({{ indicator }}, {{ area_code }}) %>%
                filter(n > 1)
        if (nrow(data) > 0) {
                area <- data %>%
                        slice(1) %>%
                        pull({{ area_code }})
                indicatorname <- data %>%
                        slice(1) %>%
                        pull({{ indicator }}) %>%
                        as.character()
                message <- paste("Some areas have multiple values for an indicator. An example is",
                                 area,
                                 "for the indicator",
                                 indicatorname)
        } else {
                message <- NA
        }
        return(message)

}

#' Data table supporting information
#'
#' Returns a data frame containing the data that sits next to the spine chart
#' @inheritParams area_profiles
#' @param dps number of decimal places to use in the data table
#' @param header_width x dimension of chart to be used for normalising the arrow
#'   length when horizonal
#' @import dplyr
#' @importFrom tidyr pivot_wider
#' @importFrom rlang  .data as_name enquo `:=`
#' @importFrom scales comma
#' @return A data frame containing the information that sits alongside the spine
#'   chart
create_datatable <- function(data, indicator,
                             area_code, timeperiod,
                             trend,
                             count, value,
                             local_area_code,
                             median_line_area_code,
                             comparator_area_code,
                             dps = 1,
                             header_width,
                             horizontal_arrow_multiplier) {
        if (is.na(comparator_area_code)) {
                area_codes <- c(local_area_code, median_line_area_code)
        } else {
                area_codes <- c(local_area_code, median_line_area_code, comparator_area_code)
        }
        data_temp <- data %>%
                filter({{ area_code }} %in% area_codes) %>%
                mutate({{ area_code }} :=
                               case_when({{ area_code }} == local_area_code ~ "Area_value",
                                         {{ area_code }} == median_line_area_code ~ "Median_value",
                                         {{ area_code }} == comparator_area_code ~ "Comparator_value",
                                         TRUE ~ "Error"),
                       {{ value }} := as.character({{ value }}),
                       dps_2b_removed = {{ dps }},
                       {{ value }} := suppressWarnings(case_when(
                               is.na(as.numeric({{ value }})) ~ {{ value }},
                               TRUE ~ ifelse(
                                       !is.na(.data$dps_2b_removed), paste0("\'",
                                                                            format(comma(round2(as.numeric({{ value }}), dps),
                                                                                         accuracy = 1 / (10 ^ dps)),
                                                                                   nsmall = 1),
                                                                            "\'"),
                                       paste0("\'",
                                              formatC(as.numeric({{ value }}),
                                                      format = "f",
                                                      big.mark = ",",
                                                      drop0trailing = TRUE),
                                              "\'")
                                       )))) %>%
                select({{ indicator }}, {{ area_code }}, {{ timeperiod }}, {{ value }}) %>%
                tidyr::pivot_wider(names_from = {{ area_code }},
                                   values_from = {{ value }})
        data_count <- data %>%
                filter({{ area_code }} == local_area_code) %>%
                select({{ indicator }}, {{ count }}) %>%
                mutate({{ count }} :=
                               suppressWarnings(ifelse(is.na(as.integer({{ count }})),
                                                       {{ count }},
                                                       paste0("'",
                                                              scales::comma(round2(as.numeric({{ count }}), 0),
                                                                            accuracy = 1),
                                                              "'"))))
        if (is.character(rlang::eval_tidy(enquo(trend), data)) |
            is.factor(rlang::eval_tidy(enquo(trend), data))) {

                data_trend <- data %>%
                        filter(({{ area_code }}) == local_area_code) %>%
                        select({{ indicator }}, {{ trend }}) %>%
                        mutate(direction = case_when(
                                grepl("decreasing", tolower({{ trend }})) ~ pi,
                                grepl("increasing", tolower({{ trend }})) ~ 0,
                                grepl("no significant change", tolower({{ trend }})) ~ pi / 2,
                                TRUE ~ NA_real_),
                               trend_sig = case_when(
                                       grepl("better", tolower({{ trend }})) ~ "Better",
                                       grepl("worse", tolower({{ trend }})) ~ "Worse",
                                       grepl("no significant change", tolower({{ trend }})) ~ "Similar",
                                       grepl("increasing", tolower({{ trend }})) ~ "Higher",
                                       grepl("decreasing", tolower({{ trend }})) ~ "Lower",
                                       TRUE ~ "Not compared"),
                               radius = case_when(
                                       grepl("no significant change", tolower({{ trend }})) ~ 0.1 * header_width * horizontal_arrow_multiplier / (n() * 1.5),
                                       grepl("increasing|decreasing", tolower({{ trend }})) ~ 0.1,
                                       TRUE ~ NA_real_
                                       )) %>%
                        select(-{{ trend }})
        } else {
                data_trend <- data %>%
                        select({{ indicator }}) %>%
                        unique() %>%
                        mutate(direction = NA,
                               trend_sig = "",
                               radius = NA)

        }
        data_temp <- merge(data_temp, data_count,
                           by = rlang::as_name(rlang::enquo(indicator)),
                           all.x = TRUE) %>%
                merge(data_trend,
                      by = rlang::as_name(rlang::enquo(indicator)),
                      all.x = TRUE) %>%
                select({{ indicator }},
                       "direction",
                       "trend_sig",
                       {{ timeperiod }},
                       {{ count }},
                       everything())


        return(data_temp)
}


#' Rescale spine data
#'
#' Rescales data so it can be plotted on a spine chart
#' @return A list containing "bars" and "points" which contains data that can be
#'   passed to the phe_spine_chart function
#' @inheritParams area_profiles
#' @inheritParams create_datatable
#' @import dplyr
#' @importFrom tibble column_to_rownames rownames_to_column
#' @importFrom purrr map map_df pmap
#' @importFrom stringr str_trim str_locate str_extract
#' @importFrom tidyr pivot_longer
#' @importFrom stats quantile
#' @importFrom scales comma
#' @importFrom rlang as_name enquo .data `:=`
spine_rescaler <- function(data,
                           area_code,
                           indicator,
                           significance,
                           polarity,
                           area_type,
                           value,
                           timeperiod,
                           local_area_code,
                           median_line_area_code,
                           comparator_area_code = NA,
                           percent_display,
                           dps = 1) {
        # make sure value field is numeric and doesn't contain annotations (based on Health Profiles annotations)
        data <- data %>%
                mutate({{ value }} :=
                               suppressWarnings(as.character({{ value }})),
                       {{ value }} :=
                               ifelse(grepl("^[0-9]", {{ value }}),
                                      str_extract({{ value }}, "^\\d+\\.*\\d*"),
                                      NA),
                       {{ value }} :=
                               suppressWarnings(as.numeric({{ value }})))

        areatype <- filter(data, {{ area_code }} == local_area_code) %>%
                pull({{ area_type }}) %>%
                unique
        remove_data <- data %>%
                filter({{ area_type }} == areatype) %>%
                group_by({{ indicator }}) %>%
                summarise(percent_na = sum(is.na({{ value }})) / n()) %>%
                filter(.data$percent_na >= percent_display) %>%
                pull({{ indicator }})

        # convert indicators to remove to na
        if (length(remove_data) > 0) {
                data <- data %>%
                        mutate({{ value }} :=
                                       ifelse({{ indicator }} %in% remove_data,
                                              NA,
                                              {{ value }}))
        }

        create_point_data <- function(data, areacode){
                if (areacode %in% c(median_line_area_code, comparator_area_code)){
                        data <- data %>%
                                filter({{ area_code }} == areacode) %>%
                                select({{ indicator }}, {{ value }}) %>%
                                rename(regionalvalue = {{ value }})
                } else {
                        data <- data %>%
                                filter({{ area_code }} == areacode) %>%
                                select({{ indicator }}, {{ significance }},
                                       {{ polarity }}, {{ value }}) %>%
                                rename(areavalue = {{ value }},
                                       Significance = {{ significance }})
                }
                data <- data.frame(data)
        }
        if (!is.na(comparator_area_code))
                parentdata <- create_point_data(data, comparator_area_code)
        areadata <- create_point_data(data, local_area_code)
        mean <- create_point_data(data, median_line_area_code)
        data <- filter(data, {{ area_type }} == areatype)

        quantiles <- data %>%
                split(pull(data, {{ indicator }})) %>%
                purrr::map(rlang::as_name(rlang::enquo(value))) %>%
                map_df(quantile, na.rm = TRUE, .id = rlang::as_name(rlang::enquo(indicator))) %>%
                data.frame()

        names(quantiles)[-1] <- c(paste0("Q", 100 * seq(0, 1, by = 0.25)))
        quantiles[, "Q50"] <- NULL
        quantiles <- quantiles %>%
                merge(mean,
                      by = rlang::as_name(rlang::enquo(indicator)),
                      all.x = TRUE) %>%
                rename(mean = "regionalvalue")

        scaled_spine_inputs <- function(IndicatorName, Q0, Q25, mean, Q75, Q100, Significance, Polarity, areavalue, regionalvalue) {
                Polarity <- stringr::str_trim(Polarity)
                quantiles <- structure(as.numeric(c(Q0, Q25, mean, Q75, Q100)),
                                       names = c("0%", "25%", "mean", "75%", "100%"))
                areavalue <- as.numeric(areavalue)
                if (!is.na(regionalvalue))
                        regionalvalue <- as.numeric(regionalvalue)
                if (grepl("Low is good",Polarity)) {
                        quantiles <- rev(quantiles)
                }
                scale_min <- ifelse(quantiles["mean"] - quantiles["0%"] >
                                            quantiles["100%"] - quantiles["mean"],
                                    quantiles["0%"],
                                    quantiles["mean"] - (quantiles["100%"] - quantiles["mean"]))
                scale_max <- ifelse(scale_min == quantiles["0%"],
                                    quantiles["mean"] + (quantiles["mean"] - quantiles["0%"]),
                                    quantiles["100%"])

                rescale <- function(val){
                        rescale <- (val - scale_min) / (scale_max - scale_min)
                        return(rescale)
                }
                quantiles <- rescale(quantiles[names(quantiles) != "mean"])
                if (!is.na(regionalvalue)) {
                        pointdata <- rescale(c(areavalue,regionalvalue))
                        names(pointdata) <- c("area","region")
                } else {
                        pointdata <- rescale(areavalue)
                        names(pointdata) <- "area"
                }

                if (grepl("Low is good",Polarity)) {
                        quantiles <- 1 - quantiles
                        quantiles <- diff(c(0,quantiles))
                        pointdata <- 1 - pointdata
                } else {
                        quantiles <- diff(c(0,quantiles))
                }

                graphpoints <- c("Worst","Q25","Q75","Best")
                scaled_spine_inputs <- list(bars = tibble({{ indicator }} := IndicatorName,
                                                          quantiles = quantiles,
                                                          GraphPoint = factor(graphpoints, levels = rev(graphpoints))),
                                            points = tibble({{ indicator }} := IndicatorName,
                                                            significance = Significance,
                                                            area = pointdata[1],
                                                            region = pointdata[2]))
        }

        dfgraph <- merge(quantiles, areadata,
                         by = rlang::as_name(rlang::enquo(indicator)),
                         all.x = TRUE)
        if (!is.na(comparator_area_code)) {
                dfgraph <- dfgraph %>%
                        merge(parentdata,
                              by = rlang::as_name(rlang::enquo(indicator)),
                              all.x =TRUE)
        } else {
                dfgraph <- dfgraph %>%
                        mutate(regionalvalue = NA)
        }

        dfgraph <- dfgraph %>%
                rename(IndicatorName = {{ indicator }},
                       Polarity = {{ polarity }}) %>%
                lapply(purrr::map, .f = as.character) %>%
                pmap(scaled_spine_inputs)
        dfgraphfinal <- list(bars = suppressWarnings(map_df(dfgraph, "bars")),
                             points = suppressWarnings(map_df(dfgraph, "points")))
        dfpolarity <- areadata %>%
                select({{ indicator }}, {{ polarity }})
        dfannotate <- quantiles %>%
                merge(dfpolarity,
                      by = rlang::as_name(rlang::enquo(indicator)),
                      all.x =TRUE) %>%
                mutate(reverse = ifelse(grepl("Low is good", {{ polarity }}),
                                        TRUE,
                                        FALSE),
                       Worst = ifelse(.data$reverse == TRUE, .data$Q100, .data$Q0),
                       Best = ifelse(.data$reverse == TRUE, .data$Q0, .data$Q100)) %>%
                select({{ indicator }}, "Best", "Worst") %>%
                tidyr::pivot_longer(names_to = "GraphPoint",
                                    values_to = "label",
                                    cols = "Best":"Worst") %>%
                mutate(y = ifelse(.data$GraphPoint == "Best", 1.05, -0.05),
                       dps_2b_removed = dps,
                       label = ifelse(is.na(.data$label),
                                      NA,
                                      ifelse(!is.na(.data$dps_2b_removed),
                                             format(comma(round2(as.numeric(.data$label), dps),
                                                          accuracy = 1 / (10 ^ dps)), nsmall = 1),
                                             formatC(as.numeric(.data$label),
                                                     format = "f",
                                                     big.mark = ",",
                                                     drop0trailing = TRUE))),
                       GraphPoint = factor(.data$GraphPoint, levels = c("Best","Q75","Q25","Worst"))) %>%
                select(!c("dps_2b_removed"))

        timeperiod <- data %>%
                select({{ indicator }}, {{ timeperiod }}) %>%
                ungroup %>%
                unique %>%
                mutate({{ indicator }} := as.character({{ indicator }}))

        areadata <- areadata %>%
                select({{ indicator }}, "areavalue")
        mean <- mean %>%
                rename(England = "regionalvalue")

        if (!is.na(comparator_area_code))
                mean <- merge(parentdata, mean,
                              by = rlang::as_name(rlang::enquo(indicator)),
                              all =TRUE)
        dfannotatepoints <- merge(mean, areadata,
                                  by = rlang::as_name(rlang::enquo(indicator)),
                                  all =TRUE) %>%
                merge(timeperiod,
                      by = rlang::as_name(rlang::enquo(indicator)),
                      all = TRUE) %>%
                mutate(England = as.character(.data$England),
                       England = case_when(
                        is.na(as.numeric(.data$England)) ~ .data$England,
                        TRUE ~ format(round2(as.numeric(.data$England), dps), nsmall = 1)),
                       areavalue = as.character(.data$areavalue),
                       areavalue = case_when(
                               is.na(as.numeric(.data$areavalue)) ~ .data$areavalue,
                               TRUE ~ format(round2(as.numeric(.data$areavalue), dps), nsmall = 1)))
        if (!is.na(comparator_area_code))
                dfannotatepoints <- dfannotatepoints %>%
                mutate(regionalvalue = as.character(.data$regionalvalue),
                       regionalvalue = case_when(
                        is.na(as.numeric(.data$regionalvalue)) ~ .data$regionalvalue,
                        TRUE ~ format(round2(as.numeric(.data$regionalvalue), dps), nsmall = 1)))

        dfgraphfinal$bars <- merge(dfgraphfinal$bars,
                                       dfannotate,
                                       by = c(rlang::as_name(rlang::enquo(indicator)), "GraphPoint"),
                                       all.x =TRUE)
        dfgraphfinal$points <- merge(dfgraphfinal$points,
                                         dfannotatepoints,
                                         by = rlang::as_name(rlang::enquo(indicator)),
                                         all.x =TRUE) %>%
                rename({{ significance }} := significance)
        if (is.na(comparator_area_code)) dfgraphfinal$points$region <- NULL
        return(dfgraphfinal)
}

#' Proper rounding of values
#' @param val numeric value to round
#' @param dps numeric, number of decimal places
#' @details function taken from this link (\url{https://stackoverflow.com/questions/12688717/round-up-from-5})
round2 <- function(val, dps) {
        posneg = sign(val)
        z = abs(val)*10^dps
        z = z + 0.5
        z = trunc(z)
        z = z/10^dps
        z*posneg
}
